### PR TITLE
Rest High Level client: Add List Tasks

### PR DESCRIPTION
**Pull request by Van0SS**
Related to #27205

### CODE REVIEW

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @Van0SS, thanks for your help on the client. This PR aready looks quiet good, I left a couple of questions I had while reviewing it, especially around the need to implement equals/hashCode for some response and failure classes. If we want to keep that then some of the tests might need some updates to test those more thorougly. But as far as the rest goes I'm okay with most of it. 
If you could merge in master and fix the two long lines, then I can also kick off another CI test run while working on the rest, so we can see if all test pass.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java
Comment (Van0SS) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java]: To parse the response we only need `group_by=none`. Grouping by parent or nodes will be done on the client side when you call `getTaskGroups()` or `getPerNodeTasks()` in `ListTasksResponse`.
But, by default API returns `group_by=parent` result, so I added here specific parameter.

File: server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java]: I did not find directly usages of `TaskOperationFailure::toString` and for me seems that nobody should rely on `toString` until it's specifically mentioned in JavaDoc for this method, maybe I live in a parallel universe :)

File: server/src/main/java/org/elasticsearch/action/support/tasks/BaseTasksResponse.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/action/support/tasks/BaseTasksResponse.java]: Correct

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java]: Parser for `ListTasksResponse` uses only `groupby=node` response format, so better `toXContent` also serialize to `groupby=node` format. It's not necessarily but a nice thing as it seems for me.

**Comment by Van0SS** [comment]
@cbuescher thanks for the review! I fixed checkstyle and merged to master, you can try to run build again.

File: server/src/test/java/org/elasticsearch/tasks/TaskInfoTests.java
Comment (Van0SS) [reply server/src/test/java/org/elasticsearch/tasks/TaskInfoTests.java]: Agree will add `mutateInstance`

**Comment by Van0SS** [comment]
@cbuescher  I am also on the fence about equals/hashcode for Classes which have Exceptions. Will wait what do you guys decide. If decision will be to use them then I will added mutations to tests.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: My mistake, they should be here.

**Comment by cbuescher** [comment]
@elasticsearch test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java
Comment (cbuescher) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/Request.java]: Thanks for clearing this up for me, makes sense.

File: server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java]: No, I agree. Nevertheless we might want to play it save, but I will get another opinion on this.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java]: So this changes the structure of the Rest response output? In this case we shouldn't do it for backward compatiblity reasons. This change should also be merged to the 6.x branches I think.

**Comment by cbuescher** [comment]
@Van0SS thanks a lot, looks great. I left some comments, will try to get some feedback on the open questions. In the meantime I kicked off some full test runs, although the PR might still change but just in case this might catch anything unexpected.

**Comment by cbuescher** [comment]
Never mind the failing packaging tests, I think that's a temporary issue we have since we updated version constants on master a few hours ago. I should resolve itself once merging master again.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java]: Yes, for the requests coming though the client I see this, what about requests that set `groupby` differently themselves? I see you also changes ListTaskResponse#toXContent to use toXContentGroupedByNone instead of toXContentGroupedByParents, I think thats why this switch here doesn't change anything but I'd still like to understand the reason for that change better.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: I overlooked this in my first round of reviews, I'd like to undestand why this needs changing.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestListTasksAction.java]: I am sorry, I thought all this conversation was about not only this 2 lines of code switching 2 parameters in `if` clause, but about all the changes including `ListTaskResponse#toXContent` to use `toXContentGroupedByNone` instead of `toXContentGroupedByParents`.
So all these changes doesn't change REST endpoint behavior. It just change `toXContent` result to be `groupby=none`, it's not necessarily but can bring:
- Idiomatically correct serialization/deserialization workflow: `toXContent` returns content which can be deserialized in `fromXContent`
- Tests can use Serialization handlers like `XContentHelper::toXContent` or can be fully automated i.e. extend `AbstractStreamableXContentTestCase` or `AbstractSerializingTestCase` (Currently not applicable because of broken `equals`)

And again it just a cosmetic change and I am happy to revert it back :)

File: server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java]: I think that this change is not necessary, and changing it is a risk. It could just be that this is logged somewhere and the current toString output makes sense the way it is, maybe even more readable than the json output.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: as far as I understand this makes parsing easier, yet there is the concern that it may break existing transport client users calling `toXContent` on the obtained response object, given that the method is public. I wonder if we could add parsing without this slight breakage and how much work that would be. It is odd that we need this behaviour change technically only  to make our tests work, given that `fromXContent` knows to parse only the `groupedByNone` output. What options do we have?

- make fromXContent parse the groupedByParents output instead?
- make it possible in some way to call toXContentGroupedByNone in our tests without changing the default behaviour of toXContent. This may be possible by making the output depend on a param passed using the params argument. The test would also have to set that param, which is not supported at the moment. This sounds rather complicated, not sure it' worth it
- Accepting this slight breaking change

Other alternatives?

**Review by cbuescher — CHANGES_REQUESTED** [review]
@Van0SS sorry for the delay, I finally got around asking @javanna for another opinion on the open topics, we agreed on the following:

* prefer to leave "TaskOperationFailure#toString" as is, since its not strictly needed for this PR and it might be used in logging somewhere where the json format is not ideal
* not implementing equals/hashCode for TaskOperationFailure and ListTaskResponse for the reasons stated. Also that means we don't need to test it, but you might need to find a way to work around the use of equals() for serialization/parsing tests in the base tests. See comments and let me know if there are any questions
* need to investigate the options for changing the ListTaskResponse#toXContent output. While I prefer the change the way you do it, there is a marginal possibility that some user relies on the current behaviour, making this a breaking Java change. This might be acceptable, but maybe there are alternatives.

Let me know what you think, again other than that this is very close to be ready.

**Comment by cbuescher** [comment]
It also looks like there are some changes that need to be merged in from master.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: @javanna I hear you, if non-breaking changes is that important then better find solution without it.

Another alternative - don't change default serialization and use custom testcase. Like it is now in this PR in `ListTasksResponseTests::testXContentWithFailures()`.
So, in this testcase we can explicitly call `ListTasksResponse::toXContentGroupedByNone`, parse, do object comparison, etc.
Not the best solution also though.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: Could we move toXContent to look at params and call the proper method depending on that? Could the REST action call that instead of the custom rest builder listeners? Then we could reuse your test changes from the cluster health PR, where you can define ToXContent.Params through a new protected method and provide the one that we need in the test? Not sure if I am missing something, I haven't tried this out but this could be the way to go.

**Comment by javanna** [comment]
hi @Van0SS would you have time to address the latest review comments? Also you'll need to merge master in and fix conflicts, for the most part `Request` is now called `RequestConverters`. Let us know if we can help.

**Comment by Van0SS** [comment]
@javanna Sorry didn't have time recently, will try to make changes soon. Thanks for ideas!

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: @javanna It is nice idea, but I stuck with implementation - `toXContentGroupedByNode` requires `DiscoveryNodes` parameter, which cannot be passed as `Params` bc `Params` can store only `String` values.
We can add `paramAsObject` and add implementation which can store any `Object` but not sure that it is a good idea. What do you think?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: I see the problem @Van0SS thanks for bringing that up. I am thinking that it's not wort to refactor this for the corner case that we are worrying about.

**Comment by cbuescher** [comment]
@Van0SS I took the liberty and merged in master to see what is left to be done. I will go over the PR once again and see if there are any other mayor things missing now that @javanna is okay with the ListTasksResponse toXContent change.

File: server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/TaskOperationFailure.java]: Related to the former discussion, would you change this back please?

**Review by cbuescher — CHANGES_REQUESTED** [review]
Currently the main things to change here are removing the mentioned equals/hashcode implementations and reverting the changes in the toString() method of TaskOperationFailure. Also I think there was some discussion about extending some tests adding the mutate function(). 
I will still kick off a CI run to see if the rest looks okay already.

**Comment by cbuescher** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java
Comment (Van0SS) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: @javanna Absolutely :)

**Comment by Van0SS** [comment]
@cbuescher Thanks for merging! Working on final changes.

**Comment by cbuescher** [comment]
@Van0SS thanks for the update, will take a look at it soon, just kicking of another test run in the meantime
@elasticmachine test this please

File: server/src/test/java/org/elasticsearch/action/TaskOperationFailureTests.java
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/action/TaskOperationFailureTests.java]: I think this is still a useful test to have, just without relying on equals/hashCode. So maybe you could structure it similar to ListTasksResponseTests, e.g. extend AbstractXContentTestCase using assertEqualInstances?

**Review by cbuescher — CHANGES_REQUESTED** [review]
Looks great, I left only a few minor comments, and maybe you can resurrect TaskOperationFailureTests that you removed in the last commit in a slightly different manner that doesn't require hashCode/Equals?

**Review by cbuescher — APPROVED** [review]
LGTM, will start the tests.

**Comment by cbuescher** [comment]
@Van0SS finally a clean CI run. Thanks so much for sticking with this through all the discussion about the nitty-gritty bwc details. It's a great PR and a valuable addition to the ongoing effort to make the high level rest client more complete. Much appreciated.
I will also merge this to the current 6.x branch so this should be coming out with 6.4.

**Comment by Van0SS** [comment]
Woohoo! Thank you guys so much for your effort and patience.

**Pull request merged**
